firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleLoader.java

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.bundle;
+
+import static com.google.firebase.firestore.model.DocumentCollections.emptyMaybeDocumentMap;
+
+import androidx.annotation.Nullable;
+import com.google.firebase.database.collection.ImmutableSortedMap;
+import com.google.firebase.database.collection.ImmutableSortedSet;
+import com.google.firebase.firestore.LoadBundleTaskProgress;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.MaybeDocument;
+import com.google.firebase.firestore.model.NoDocument;
+import com.google.firebase.firestore.util.Preconditions;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A class to process the elements from a bundle, load them into local storage and provide progress
+ * update while loading.
+ */
+public class BundleLoader {
+  private final BundleListener bundleListener;
+  private final BundleMetadata bundleMetadata;
+  private final int totalDocuments;
+  private final long totalBytes;
+  private final List<NamedQuery> queries;
+  private final Map<DocumentKey, BundledDocumentMetadata> documentsMetadata;
+
+  private ImmutableSortedMap<DocumentKey, MaybeDocument> documents;
+  private long bytesLoaded;
+  @Nullable private DocumentKey currentDocument;
+
+  public BundleLoader(
+      BundleListener bundleListener,
+      BundleMetadata bundleMetadata,
+      int totalDocuments,
+      long totalBytes) {
+    this.bundleListener = bundleListener;
+    this.bundleMetadata = bundleMetadata;
+    this.totalDocuments = totalDocuments;
+    this.totalBytes = totalBytes;
+    this.queries = new ArrayList<>();
+    this.documents = emptyMaybeDocumentMap();
+    this.documentsMetadata = new HashMap<>();
+  }
+
+  /**
+   * Adds an element from the bundle to the loader.
+   *
+   * <p>Returns a new progress if adding the element leads to a new progress, otherwise returns
+   * null.
+   */
+  public @Nullable LoadBundleTaskProgress addElement(BundleElement bundleElement, long byteSize) {
+    Preconditions.checkArgument(
+        !(bundleElement instanceof BundleMetadata), "Unexpected bundle metadata element.");
+
+    boolean updateProgress = false;
+
+    if (bundleElement instanceof NamedQuery) {
+      queries.add((NamedQuery) bundleElement);
+    } else if (bundleElement instanceof BundledDocumentMetadata) {
```

Comment:
I am wondering if we could bring the implementation more consistent to the JS implementation, or maybe there is something I don't see that is blocking it?

This is mostly a concern of inconsistency, but JS also does not have to use a sorted map.

Reply:
There are two differences here that stand out to me:

- The metadata lookup in JS looks at the gathered documents. I decided to not follow this pattern as I would like to build up an `ImmutableSortedMap<DocumentKey, MaybeDocument>` directly as this is a common data structure I can use this map directly in LocalStore. As such, I need some place to denote the last stored document key. 
I can remove this altogether though since it is only used for asserts.

- The `updateProgress` check is done differently here. I prefer the `updateProgress` a little, but not enough to convince you that it is worth backporting. I updated it to be more aligned with the Web.

Reply:
Looks good.